File: C++/Concurrency/RaceDemo.cpp
#include <atomic>
#include <condition_variable>
#include <functional>
#include <iostream>
#include <mutex>
#include <queue>
#include <thread>
#include <vector>

class ThreadPool {
public:
    explicit ThreadPool(size_t numThreads) {
        for (size_t i = 0; i < numThreads; ++i) {
            workers_.emplace_back([this] { workerLoop(); });
        }
    }

    ~ThreadPool() {
        {
            std::unique_lock<std::mutex> lock(mutex_);
            stop_ = true;
        }
        cond_.notify_all();
        for (auto& worker : workers_) worker.join();
    }

    void submit(std::function<void()> task) {
        {
            std::unique_lock<std::mutex> lock(mutex_);
            tasks_.push(std::move(task));
        }
        cond_.notify_one();
    }

private:
    void workerLoop() {
        while (true) {
            std::function<void()> task;
            {
                std::unique_lock<std::mutex> lock(mutex_);
                cond_.wait(lock, [this] { return stop_ || !tasks_.empty(); });
                if (stop_ && tasks_.empty()) return;
                task = std::move(tasks_.front());
                tasks_.pop();
            }
            task();
        }
    }

    std::vector<std::thread> workers_;
    std::queue<std::function<void()>> tasks_;
    std::mutex mutex_;
    std::condition_variable cond_;
    bool stop_ = false;
};

int unsafeCounter = 0;
std::atomic<int> safeCounter{0};

int main() {
    {
        ThreadPool pool(8);
        for (int i = 0; i < 1000; ++i) {
            pool.submit([] {
                unsafeCounter++;           // NOT atomic — data race
                safeCounter.fetch_add(1);  // atomic
            });
        }
    }  // pool destructor joins all workers before continuing

    std::cout << "unsafe: " << unsafeCounter << std::endl;      // almost never 1000
    std::cout << "safe:   " << safeCounter.load() << std::endl;  // always 1000
    return 0;
}

File: Golang/Concurrency/RaceDemo.go
package main

import (
	"fmt"
	"sync"
	"sync/atomic"
)

func main() {
	var unsafeCounter int
	var safeCounter int64

	tasks := make(chan struct{}, 1000)
	for i := 0; i < 1000; i++ {
		tasks <- struct{}{}
	}
	close(tasks)

	var wg sync.WaitGroup
	for i := 0; i < 8; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for range tasks {
				unsafeCounter++                  // NOT atomic — data race
				atomic.AddInt64(&safeCounter, 1) // atomic
			}
		}()
	}
	wg.Wait()

	fmt.Println("unsafe:", unsafeCounter)                   // almost never 1000
	fmt.Println("safe:  ", atomic.LoadInt64(&safeCounter)) // always 1000
}
